app/src/main/java/me/hufman/androidautoidrive/phoneui/SpotifyAuthorizationActivity.kt

```diff
@@ -0,0 +1,240 @@
+ */
+class SpotifyAuthorizationActivity: Activity() {
+	companion object {
+		const val TAG = "AuthorizationActivity"
+
+		const val REQUEST_CODE_SPOTIFY_LOGIN = 2356
+
+		const val EXTRA_AUTHORIZATION_RESULT = "authorizationResult"
+		const val AUTHORIZATION_CANCELED = 1
+		const val AUTHORIZATION_SUCCESS = 0
+		const val AUTHORIZATION_FAILED = -1
+
+		val CODE_VERIFIER = generateCodeVerifierString()
+
+		/**
+		 * Generates a 45 character random alpha numeric string to be used for the code verifier in
+		 * the PKCE authentication flow.
+		 */
+		private fun generateCodeVerifierString(): String {
+			val charPool = ('a'..'z') + ('A'..'Z') + ('0'..'9')
+			return (1..45).map { charPool[Random.nextInt(0, charPool.size)] }.joinToString("")
+		}
+	}
+
+	private val authRequest = AtomicReference<AuthorizationRequest>()
+	private val authIntent = AtomicReference<CustomTabsIntent>()
+	private lateinit var authService: AuthorizationService
+	private var authIntentLatch = CountDownLatch(1)
+	private lateinit var authStateManager: SpotifyAuthStateManager
+	private lateinit var appSettingsReceiver: MutableAppSettingsReceiver
+	private val scopes = listOf(
+			SpotifyScope.USER_MODIFY_PLAYBACK_STATE.uri,
+			SpotifyScope.USER_LIBRARY_READ.uri
+	)
+
+	/**
+	 * Performs the authorization request
+	 */
+	private fun doAuth() {
+		Log.d(TAG, "Authorization started")
+		try {
+			authIntentLatch.await()
+		} catch (ex: InterruptedException) {
+			Log.w(TAG, "Interrupted while waiting for auth intent")
+		}
+
+		val intent = authService.getAuthorizationRequestIntent(
+				authRequest.get(),
+				authIntent.get())
+
+		startActivityForResult(intent, REQUEST_CODE_SPOTIFY_LOGIN)
+	}
+
+	private fun warmUpBrowser() {
+		authIntentLatch = CountDownLatch(1)
+		Log.d(TAG, "Warming up browser instance for auth request")
+		val intentBuilder = authService.createCustomTabsIntentBuilder(authRequest.get().toUri())
+		authIntent.set(intentBuilder.build())
+		authIntentLatch.countDown()
+	}
+
+	private fun createAuthRequest() {
+		Log.d(TAG, "Creating auth request")
+		val codeChallenge = getSpotifyPkceCodeChallenge(CODE_VERIFIER)
+		val codeChallengeMethod = "S256"
+		val clientId = packageManager.getApplicationInfo(packageName, PackageManager.GET_META_DATA)
+				.metaData.getString("com.spotify.music.API_KEY", "unavailable")
+		val authRequestBuilder = AuthorizationRequest.Builder(
+				authStateManager.getAuthorizationServiceConfiguration()!!,
+				clientId,
+				ResponseTypeValues.CODE,
+				Uri.parse(SpotifyAppController.REDIRECT_URI))
+				.setScopes(scopes)
+				.setCodeVerifier(CODE_VERIFIER, codeChallenge, codeChallengeMethod)
+		authRequest.set(authRequestBuilder.build())
+	}
+
+	private fun initializeAppAuth() {
+		Log.d(TAG, "Initializing AppAuth")
+		recreateAuthorizationService()
+
+		if (authStateManager.getAuthorizationServiceConfiguration() == null) {
+			val authEndpointUri = Uri.parse("https://accounts.spotify.com/authorize")
```

Comment:
These seem like good constants to pull up to the `companion object`

Reply:
Yeah that's probably a better place for those, moved those URI strings up to the companion object